File: client/contacts/QueryLanguageExample.cs
/*
 * #%L
 * QueryLanguageExample.cs
 * ---
 * Copyright (C) 2000 - 2017 Oracle and/or its affiliates. All rights reserved.
 * ---
 * Oracle is a registered trademarks of Oracle Corporation and/or its
 * affiliates.
 * 
 * This software is the confidential and proprietary information of Oracle
 * Corporation. You shall not disclose such confidential and proprietary
 * information and shall use it only in accordance with the terms of the
 * license agreement you entered into with Oracle.
 * 
 * This notice may not be removed or altered.
 * #L%
 */
﻿
using System;
using System.Collections;
using Tangosol.Net;
using Tangosol.Util;
using Tangosol.Util.Aggregator;

namespace Tangosol.Examples.Contacts    
{
    /// <summary>
    /// QueryLanguageExample runs sample queries for contacts.
    /// </summary>
    /// <remarks>
    /// <p>
    /// The purpose of this example is to show how to create a
    /// ReflectionExtrector on cache data and how to create a KeyExtractor
    /// index for the cache keys.</p>
    /// It also illustrates how to utilize the indexes to filter the data set
    /// how to efficiently create a matching set.<p/>
    /// <p>
    /// Finally the example demonstrates how to use some of the built-in
    /// cache aggregators to do simple computational tasks on the cache data.</p>
    /// </remarks>
    /// <author>Ivan Cikic  2010.03.09</author>
    public class QueryLanguageExample
    {
        /// <summary>
        /// Create indexes in the cache and query it for data.
        /// </summary>
        /// <param name="cache">
        /// Cache to query.
        /// </param>
        /// <param name="ff">
        /// The <see cref="FilterFactory"/> used to convert string to
        /// <b>IFilters</b>.
        /// </param>
        public void Query(INamedCache cache, FilterFactory ff)
        {
            Console.WriteLine("------QueryLanguageExample begins------");

            // Add indexes to make queries more efficient
            // Ordered index applied to fields used in range and like filter queries
            cache.AddIndex(ff.CreateExtractor("age"), /*fOrdered*/ true, /*comparator*/ null);
            cache.AddIndex(ff.CreateExtractor("key().lastName"), /*fOrdered*/ true, /*comparator*/ null);
            cache.AddIndex(ff.CreateExtractor("homeAddress.city"), /*fOrdered*/ true, /*comparator*/ null);
            cache.AddIndex(ff.CreateExtractor("homeAddress.state"), /*fOrdered*/ false, /*comparator*/ null);
            cache.AddIndex(ff.CreateExtractor("workAddress.state"), /*fOrdered*/ false, /*comparator*/ null);

            // Find all contacts who live in Massachusetts
            ICollection results = cache.GetEntries(ff.CreateFilter("homeAddress.state = 'MA'"));
            PrintResults("MA Residents", results);

            // Find all contacts who live in Massachusetts and work elsewhere
            results = cache.GetEntries(ff.CreateFilter("homeAddress.state is 'MA' and workAddress is not 'MA'"));
            PrintResults("MA Residents, Work Elsewhere", results);

            // Find all contacts whose city name begins with 'S'
            results = cache.GetEntries(ff.CreateFilter("homeAddress.city like 'S%'"));
            PrintResults("City Begins with S", results);

            const int age = 58;
            object[]  env =  new object[] { age };
            // Find all contacts who are older than nAge
            results = cache.GetEntries(ff.CreateFilter("age > ?1", env));
            PrintResults("Age > " + age, results);

            // Find all contacts with last name beginning with 'S' that live
            // in Massachusetts. Uses both key and value in the query
            results = cache.GetEntries(ff.CreateFilter("lastName like 'S%' and homeAddress.state = 'MA'"));
            PrintResults("Last Name Begins with S and State Is MA", results);
            // Count contacts who are older than nAge for the entire cache dataset.
            Console.WriteLine("count > {0} : {1}", age, 
                cache.Aggregate(ff.CreateFilter("age > ?1", env), new Count()));

            // Find minimum age for the entire cache dataset.
            IFilter always = ff.CreateFilter("true");
            Console.WriteLine("min age: {0}", cache.Aggregate(always, new LongMin("getAge")));

            // Calculate average age for the entire cache dataset.
            Console.WriteLine("avg age: {0}", cache.Aggregate(always, new DoubleAverage("getAge")));

            // Find maximum age for the entire cache dataset.
            Console.WriteLine("max age: {0}", cache.Aggregate(always, new LongMax("getAge")));

            Console.WriteLine("------QueryLanguageExample completed------");
        }

        /// <summary>
        /// Print results of the query.
        /// </summary>
        /// <param name="title">The title that describes the results</param>
        /// <param name="results">A set of query results.</param>
        private static void PrintResults(string title, ICollection results)
        {
            Console.WriteLine(title);
            foreach (object result in results)
            {
                Console.WriteLine(result);
            }
        }
    }
}

File: server/src/com/tangosol/examples/contacts/QueryLanguageExample.java
/*
 * #%L
 * QueryLanguageExample.java
 * ---
 * Copyright (C) 2000 - 2017 Oracle and/or its affiliates. All rights reserved.
 * ---
 * Oracle is a registered trademarks of Oracle Corporation and/or its
 * affiliates.
 * 
 * This software is the confidential and proprietary information of Oracle
 * Corporation. You shall not disclose such confidential and proprietary
 * information and shall use it only in accordance with the terms of the
 * license agreement you entered into with Oracle.
 * 
 * This notice may not be removed or altered.
 * #L%
 */
package com.tangosol.examples.contacts;

import com.tangosol.examples.pof.Contact;
import com.tangosol.examples.pof.ContactId;
import com.tangosol.net.NamedCache;

import com.tangosol.util.aggregator.Count;
import com.tangosol.util.aggregator.DoubleAverage;
import com.tangosol.util.aggregator.LongMax;
import com.tangosol.util.aggregator.LongMin;

import com.tangosol.util.Filter;

import java.util.Iterator;
import java.util.Set;

import static com.tangosol.examples.contacts.ExamplesHelper.logHeader;

/**
 * QueryLanguageExample runs sample queries for contacts.
 * <p/>
 * The purpose of this example is to show how to create Extractors and
 * Filters by using the Coherence QueryHelper interface indirectly through
 * a FilterFactory. This example does the same operations as the
 * QueryExample in the contacts example.
 *
 * @author djl 2010.03.01
 */
public class QueryLanguageExample
    {
    // ----- QueryLanguageExample methods ---------------------------------------

    /**
     * Create indexes in the cache and query it for data.
     *
     * @param cache cache to query
     * @param ff    the FilterFactory to use for making Filters
                    and ValueExtractors
     */
     @SuppressWarnings("unchecked")
     public void query(NamedCache<ContactId, Contact> cache, FilterFactory ff)
         {
         logHeader("QueryLanguageExample begins");

         // Add indexes to make queries more efficient
         // Ordered index applied to fields used in range and like filter queries
         cache.addIndex(ff.createExtractor("age"), /*fOrdered*/ true,
                 /*comparator*/ null);
         cache.addIndex(ff.createExtractor("key().lastName"),
                  /*fOrdered*/ true, /*comparator*/ null);
         cache.addIndex(ff.createExtractor("homeAddress.city"),
                 /*fOrdered*/ true, /*comparator*/ null);
         cache.addIndex(ff.createExtractor("homeAddress.state"),
                 /*fOrdered*/ false, /*comparator*/ null);
         cache.addIndex(ff.createExtractor("workAddress.state"),
                 /*fOrdered*/ false, /*comparator*/ null);

         // Find all contacts who live in Massachusetts
         Set setResults = cache.entrySet(ff.createFilter("homeAddress.state = 'MA'"));
         printResults("MA Residents", setResults);

         // Find all contacts who live in Massachusetts and work elsewhere
         setResults = cache.entrySet(
              ff.createFilter("homeAddress.state is 'MA' and workAddress is not 'MA'"));
         printResults("MA Residents, Work Elsewhere", setResults);

         // Find all contacts whose city name begins with 'S'
         setResults = cache.entrySet(ff.createFilter("homeAddress.city like 'S%'"));
         printResults("City Begins with S", setResults);

         final int nAge = 58;
         Object[] aEnv  =  new Object[] {new Integer(nAge)};
         // Find all contacts who are older than nAge
         setResults = cache.entrySet(ff.createFilter("age > ?1",aEnv));
         printResults("Age > " + nAge, setResults);

         // Find all contacts with last name beginning with 'S' that live
         // in Massachusetts. Uses both key and value in the query
         setResults = cache.entrySet(
                  ff.createFilter("lastName like 'S%' and homeAddress.state = 'MA'"));
         printResults("Last Name Begins with S and State Is MA", setResults);

         // Count contacts who are older than nAge for the entire cache dataset.
         System.out.println("count > " + nAge + ": " + cache.aggregate(
                 ff.createFilter("age > ?1", aEnv), new Count<ContactId, Contact>()));

         // Find minimum age for the entire cache dataset.
         Filter always = ff.createFilter("true");
         System.out.println("min age: " + cache.aggregate(always, new LongMin<Integer>("getAge")));

         // Calculate average age for the entire cache dataset.
         System.out.println("avg age: " + cache.aggregate(always, new DoubleAverage<Integer>("getAge")));

         // Find maximum age for the entire cache dataset.
         System.out.println("max age: " + cache.aggregate(always, new LongMax<Integer>("getAge")));

	     logHeader("QueryLanguageExample completed");
         }

    /**
     * Print results of the query
     *
     * @param sTitle     the title that describes the results
     *
     * @param setResults a set of query results
     */
    private void printResults(String sTitle, Set setResults)
        {
        System.out.println(sTitle);
        for (Iterator iter = setResults.iterator(); iter.hasNext();)
            {
            System.out.println(iter.next());
            }
        }
    }
